larq/optimizers_v2.py

```diff
@@ -0,0 +1,83 @@
+import tensorflow as tf
+import numpy as np
+import larq as lq
+
+from larq import utils
+from copy import deepcopy
+
+
+@utils.register_keras_custom_object
+class Bop(tf.keras.optimizers.Optimizer):
+    def __init__(self, fp_optimizer, threshold=1e-5, gamma=1e-2, name="Bop", **kwargs):
+        super().__init__(name=name, **kwargs)
+
+        if not isinstance(fp_optimizer, tf.keras.optimizers.Optimizer):
+            raise TypeError(
+                f"Expected tf.keras.optimizers.Optimizer, received {type(fp_optimizer)}."
+            )
+
+        self.fp_optimizer = fp_optimizer
+        self._set_hyper("threshold", threshold)
+        self._set_hyper("gamma", gamma)
+
+    def _create_slots(self, var_list):
+        for var in var_list:
+            if self.is_binary(var):
+                self.add_slot(var, "m")
+
+    def apply_gradients(self, grads_and_vars, name=None):
+        bin_grads_and_vars = [(g, v) for g, v in grads_and_vars if self.is_binary(v)]
+        fp_grads_and_vars = [(g, v) for g, v in grads_and_vars if not self.is_binary(v)]
```

Comment:
I think we should also allow people to manually pass in the binarized parameters, similar to how they do it [here](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/weight_decay_optimizers.py#L131) so people can use this optimizer with a custom training loop and custom layers. I don't think we should add it in this PR though, just some thing to keep in mind for future improvements.

Reply:
@koenhelwegen